### PR TITLE
Added discount abstractions and tests

### LINKED ISSUE #310: Discount Subsystem: Discount Abstraction
Rather than codifying the discount logic in a complicated computation function, discount types will be implemented by abstracting the logic around discounts into a registry-driven discount factory like this:

```python
import abc
from dataclasses import dataclass

@dataclass
class DiscountType(abc.ABC):
    _CLASSES = {}

    discount: Discount
    
    # see https://www.python.org/dev/peps/pep-0487/
    def __init_subclass__(cls, *, discount_type, **kwargs):
        super().__init_subclass__(**kwargs)

        if discount_type in _CLASSES:
            raise TypeError(f"{discount_type} already defined for DiscountType")

        cls.discount_type = discount_type
        cls._CLASSES[discount_type] = cls
        
    @classmethod
    def for_discount(cls, discount: Discount):
        DiscountTypeCls = cls._CLASSES[discount.discount_type]

        return DiscountTypeCls(discount)

    def get_product_price(self, product: Product):
        return self.get_product_version_price(product.latest_version)

    @abc.abstractmethod
    def get_product_version_price(self, product_version: ProductVersion): 
          ...


class PercentDiscount(DiscountType, 
          discount_type=Discount.PERCENT_DISCOUNT):

    def get_product_version_price(self, product_version: ProductVersion):
            return product_version.price * self.discount.amount

class FixedPriceDiscount(DiscountType, 
          discount_type=Discount.PERCENT_DISCOUNT):

    def get_product_version_price(self, product_version: ProductVersion):
            return self.discount.amount  # the amount here is the fixed price

```

With this implementation, prices before ordering would use `get_product_price`, whereas the receipt service would use `get_product_version_price` on the purchased versions. This makes it far more scalable to introduce new discount types without having to refactor existing code.

### PR DESCRIPTION
**Pull request by jkachel**
added discount abstraction API and tests for those

#### Pre-Flight checklist

- [X] Testing
  - [X] Code is tested

#### What are the relevant tickets?
Fixes #310 

#### What's this PR do?
Adds the discount calculation code as an abstract base class, and implementations for four discount types. 

#### How should this be manually tested?
pytest should pass

### CODE REVIEW

File: ecommerce/discounts_test.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        what_type = (
+            type(discount_logic) is DollarsOffDiscount,
+            type(discount_logic) is FixedPriceDiscount,
+            type(discount_logic) is PercentDiscount,
+        )
+
+        assert any(what_type)
+
+
+def test_discout_factory_adjustment(discounts, products):
```
Comment (annagav) [inline ecommerce/discounts_test.py:38]: typo here

File: ecommerce/discounts_test.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+def products():
+    return ProductFactory.create_batch(5)
+
+
+@pytest.fixture()
+def discounts():
+    return DiscountFactory.create_batch(10)
+
+
+def test_discount_factory_generation(discounts):
```
Comment (annagav) [inline ecommerce/discounts_test.py:25]: Could you please add comments to each test, mentioning what thei are testing?

File: ecommerce/discounts_test.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            type(discount_logic) is DollarsOffDiscount,
+            type(discount_logic) is FixedPriceDiscount,
+            type(discount_logic) is PercentDiscount,
+        )
+
+        assert any(what_type)
+
+
+def test_discout_factory_adjustment(discounts, products):
+    # product = products[random.randrange(0, len(products), 1)]
```
Comment (annagav) [inline ecommerce/discounts_test.py:39]: was this for testing purposes?

**Review by annagav — CHANGES_REQUESTED** [review]
Just couple of minor comments

File: ecommerce/discounts_test.py (hunk 1 above)
Comment (jkachel) [reply ecommerce/discounts_test.py]: Fixed in 20db5b212d945e1d53af7a820d3986145a0a8927

File: ecommerce/discounts_test.py (hunk 2 above)
Comment (jkachel) [reply ecommerce/discounts_test.py:25]: Added these in 20db5b212d945e1d53af7a820d3986145a0a8927

File: ecommerce/discounts_test.py (hunk 3 above)
Comment (jkachel) [reply ecommerce/discounts_test.py]: Initially I had this just do one product from the factory; I decided later to just run through all of them that were being generated and forgot to remove this. It's gone in 20db5b212d945e1d53af7a820d3986145a0a8927

**Review by annagav — APPROVED** [review]
LGTM

**Pull request merged**